### ISSUE TITLE
🐞 Bug Report: Post type override on autosave (ED-8591)

### ISSUE THREAD

**Issue by aporiadin** [elementor/elementor#19941]
### Prerequisites

- [X] I have searched for similar issues in both open and closed tickets and cannot find a duplicate.
- [X] The issue still exists against the latest stable version of Elementor.

### Description

The plugin overrides a document type on autosave.

It tries to set the document to a custom type when elementor pro is disabled, and the "error 404" type when elementor pro is enabled.

**Possible solution**
Make sure we don't get a cached version of the document if someone tries to get it in some WP hook.

Bypass cache here:
https://github.com/elementor/elementor/blob/master/core/base/document.php#L528

`$document = Plugin::$instance->documents->get( $autosave_id );`
->
`$document = Plugin::$instance->documents->get( $autosave_id, false );`

### Steps to reproduce

-1-
https://github.com/elementor/elementor/blob/master/core/base/document.php#L516
This call triggers the "save_post" action.

-2-
In one of the functions associated with the "save_post" action, I want to get a document by the passed ID in this way.
`\Elementor\Plugin::$instance->documents->get($post_id);`

The problem is that the plugin will cache the requested document with the wrong type because we haven't copied the meta tags yet.

-3-
The plugin then copies the meta tags here
https://github.com/elementor/elementor/blob/master/core/base/document.php#L526

-4-
Here we will get a cached document with the wrong type (cached version from step #2)
https://github.com/elementor/elementor/blob/master/core/base/document.php#L528

-5-
And as a final step, we save the wrong type here and override the metadata from step #3.
https://github.com/elementor/elementor/blob/master/core/base/document.php#L529

### Isolating the problem

- [X] This bug happens with only Elementor plugin active (and Elementor Pro).
- [X] This bug happens with a Blank WordPress theme active ([Hello theme](https://wordpress.org/themes/hello-elementor/)).
- [X] I can reproduce this bug consistently using the steps above.

### System Info

    == Active Plugins ==
	Classic Editor
		Version: 1.5
		Author: WordPress Contributors

	Elementor
		Version: 3.7.6
		Author: Elementor.com

	Gravity Forms
		Version: 2.6.3.4
		Author: Gravity Forms

**Comment by aporiadin** [comment]
Hello @nicholaszein.

Do you have any estimates on this issue?

**Comment by Nevoss** [comment]
Hey @aporiadin.
Let me understand, are you trying to update the post type of a document from the "Editor" view?
can you please bring steps to reproduce from the user's point of view?

**Comment by aporiadin** [comment]
Hi @Nevoss.

I do not want to change the post type.

The plugin overrides a type on autosave. So you need to open the editor, change something and then wait for autosave. You can find all the details above. I even described how the code works and why it happens.

**Comment by StyleShit** [comment]
Hey,

How does it affect the end-user? What a user should do in order to see this bug?
Or is it a developer-only bug? If so, how do you trigger this bug? Do you register a custom hook?
I don't fully understand where this bug meets the end-user / end-developer, so in order to investigate this issue we need more information and better STR (Steps To Reproduce) _from the user point-of-view_ (whether it's a user or a developer).

Thanks!

**Comment by aporiadin** [comment]
HI, @StyleShit

You can't reproduce this problem without changes in the code. Just to emulate another plugin.

The problem occurs when the user has a plugin that registers the "save_post" hook and uses "\Elementor\Plugin::$instance->documents->get($post_id);" inside.
You also need to activate Elementor Pro. We need more types.

Summary

1. Edit a popup with the editor
2. Wait for autosave
3. It's not a popup anymore

Why

1. [This call triggers the "save_post" action](https://github.com/elementor/elementor/blob/master/core/base/document.php#L517)
2. We are going to get the new post in our handler for "save_post", and Elementor caches it. This is how "instance->documents->get" works. The problem is that the plugin will cache the requested document with the wrong type because you haven't copied the meta tags yet.
3. Then you copy meta tags (line 527)
4. Then you get a post (line 529). This is the cached document from step #2.
5. Then you want to save post type (line 530).

So basically the problem is that you need to get the document from DB (line 529) and not from cache. The document may be cached at this point by a request from an external plugin.

On a real project, we have lost several posts due to this problem. Elementor changed type to "Error 404".

**Comment by StyleShit** [comment]
And can you solve this on your end by passing `false` as a second parameter (`$from_cache`)?

```PHP
\Elementor\Plugin::$instance->documents->get( $post_id, false );
```

**Comment by StyleShit** [comment]
Hey,

Can you confirm that this PR solves the problem?
https://github.com/elementor/elementor/pull/20421

**Comment by aporiadin** [comment]
Hi,

The PR solves the problem. Thank you a lot.

**Comment by aporiadin** [comment]
Wow, ok, I understand how the proposed solution works. Looks better than my solution from above.
Thanks again.

**Comment by StyleShit** [comment]
Great!
It's probably gonna be merged into release/3.10.0

**Comment by StyleShit** [comment]
Hey,
I changed the implementation since it caused another bug that helped figure out the root cause.
Can you confirm it still solves the problem for you?

And I'll be glad if you can share the code snippet that you're using to test this (just to make sure I'm using a proper snippet on my end).

**Comment by aporiadin** [comment]
Hi,

The fix works. Thank you.

```
add_action( 'save_post', 'on_save_post', 10, 1 );
function on_save_post( $post_id ) {
    \Elementor\Plugin::$instance->documents->get( $post_id );
}
```

**Comment by nicholaszein** [comment]
Hello there!

We have great news! 🙌 

### 📢 We're happy to announce that this issue was solved in Elementor **v3.10**! 🥳 

✔️ Please feel free to check it out and update your plugin to the new version!

Cheers 🥂
